### PR TITLE
support better error messages

### PR DESCRIPTION
**Pull request by apocelipes**
For #238.

### CODE REVIEW

**Comment by apocelipes** [comment]
This is ready for review.

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
Hi, @apocelipes, thanks for the PR! I agree that go-duckdb should support better error messages. I've left one main question: How should go-duckdb handle (or even just distinguish) between duckdb errors and driver errors?

File: errors.go [hunk 1]
```diff
@@ -80,3 +81,125 @@ var (
...
+	ErrorTypeParameterNotResolved                        // parameter types could not be resolved
+	ErrorTypeParameterNotAllowed                         // parameter types not allowed
+	ErrorTypeDependency                                  // dependency
+	ErrorTypeHTTP
+	ErrorTypeMissingExtension // Thrown when an extension is used but not loaded
+	ErrorTypeAutoLoad         // Thrown when an extension is used but not loaded
+	ErrorTypeSequence
+)
+
+var exceptionPrefixMap = map[string]DuckDBErrorType{
```
[the remark this answers is not in the document]
Comment (apocelipes) [reply errors.go:132]: errorPrefixMap is a nicer one, I will change to it, thanks.

File: errors.go [hunk 2]
```diff
@@ -80,3 +81,125 @@ var (
...
+	"HTTP Error":                   ErrorTypeHTTP,
+	"Missing Extension Error":      ErrorTypeMissingExtension,
+	"Extension Autoloading Error":  ErrorTypeAutoLoad,
+	"Sequence Error":               ErrorTypeSequence,
+}
+
+type DuckDBError struct {
+	Type DuckDBErrorType
+	Msg  string
+}
```
[the remark this answers is not in the document]
Comment (apocelipes) [reply errors.go:180]: I think dirver errors and errors from db engine is actually different things. If we take a look at the [go-mysql driver](https://github.com/go-sql-driver/mysql/blob/master/packets.go#L584), they only use type `MySQLError` to represent an error comes from the MySQL engine(normaly a SQL syntax error or just some tables/fields does not exist), so we are better to make db errors and driver errors be splited types, like `GoDuckDBError` and `DriverError`(but look below).

> I am asking this because ideally go-duckdb does not panic on driver errors like the one below.
As I know, using a closed connection is a deadly failure, on Unix/Linux it always sends a bad pipe signal and the c/c++ program crashes, so panicking on this error won’t surprise the user. A driver error can come from misusing, incorrect configurations or environment problems(etc. network), even if these errors are returned, users usually cannot take other actions to effectively solve the problem except outputting logs. My personal opinion is let these errors panic is the better way.

File: errors_test.go [hunk 3]
```diff
@@ -273,3 +273,49 @@ func TestErrAPISetValue(t *testing.T) {
 	err := chunk.SetValue(1, 42, "hello")
 	testError(t, err, errAPI.Error(), columnCountErrMsg)
 }
+
+func TestGetDuckDBError(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (apocelipes) [reply errors_test.go:277]: Using real errors is a good idea, but I must admit that some errors are hard to trigger, so it might be hard to cover all corner cases.
However I will change to test with the actual exceptions, thanks for the advice.

File: errors.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (taniabogatsch) [reply errors.go:180]: These are fair points - let's stick to introducing `DuckDBError` in this PR.

File: errors_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (taniabogatsch) [reply errors_test.go:277]: Yes, I agree; many exceptions are hard to trigger. We can add a few tests to cover the error map (constraint exception, invalid input, (maybe) not implemented), and to cover `Is(err error)` and `Error`. Also, since we expose `DuckDBError`, we don't need an additional `struct` (like below) in the tests, no?
```Go
struct {
 		msg string
 		typ DuckDBErrorType
 	}
```

[the remark this answers is not in the document]
Comment (apocelipes) [reply errors_test.go:277]: That's right. I'd like to add a test that checks the actual exceptions and simplify the `TestGetDuckDBError` to focus on coner cases.

**Review by taniabogatsch — APPROVED** [review]

**Comment by taniabogatsch** [comment]
Thanks for the contribution!
